code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +164,247 @@ public void updateAccount(Account account, String name, String kind, String exte
+                        if (memberNew.getExternalId().equalsIgnoreCase(member.getExternalId()) &&
+                                memberNew.getExternalIdType().equalsIgnoreCase(member.getExternalIdType())) {
+                            if (!memberNew.getRole().equalsIgnoreCase(member.getRole())) {
+                                update.add(memberNew);
+                                deleteIgnore.add(member);
+                            } else {
+                                createIgnore.add(memberNew);
+                                deleteIgnore.add(member);
+                            }
+                        }
+                    }
+                }
+                HashSet<Member> create = (HashSet) members.clone();
+                create.removeAll(update);
+                create.removeAll(createIgnore);
+                HashSet<Member> delete = (HashSet) otherPreviosMembers.clone();
+                delete.removeAll(deleteIgnore);
+
+                Condition allMembers = DSL.falseCondition();
+                for (Member member : delete) {
+                    allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
+                            .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(member.getExternalIdType()))
+                            .and(PROJECT_MEMBER.PROJECT_ID.eq(project.getId())));
+                }
+                List<? extends ProjectMember> toDelete = create().selectFrom(PROJECT_MEMBER).where(allMembers).fetch();
+                for (ProjectMember member : toDelete) {
+                    objectProcessManager.executeStandardProcess(StandardProcess.DEACTIVATE, member, null);
+                    objectProcessManager.executeStandardProcess(StandardProcess.REMOVE, member, null);
+                }
+                List<ProjectMember> newMembers = new ArrayList<>();
+                for (Member member : create) {
+                    newMembers.add(createProjectMember(project, member));
+                }
+                for (Member member : update) {
+                    newMembers.add(updateMemberRole(project, member));
+                }
+                newMembers.addAll(getProjectMembersByMembers(createIgnore, project.getId()));
+                return newMembers;
+            }
+        });
     }
 
-    @Inject
-    public void setGenericResourceDao(GenericResourceDao resourceDao) {
-        this.resourceDao = resourceDao;
+    private List<? extends ProjectMember> getProjectMembersByMembers(Set<Member> members, long projectId) {
+        Condition allMembers = DSL.falseCondition();
+        for (Member member: members){
+            allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
+                    .and(PROJECT_MEMBER.EXTERNAL_ID_TYPE.eq(member.getExternalIdType()))
+                    .and(PROJECT_MEMBER.ROLE.in(member.getRole()))
+                    .and(PROJECT_MEMBER.PROJECT_ID.eq(projectId))
```

Comment:
We should move the checks for PROJECT_ID, STATE, and REMOVED outside of the loop since they don't change per iteration (unlike the fields within Member).  This is just an optimization.